packages/protocol/contracts/L1/libs/LibProvingAlt.sol

```diff
@@ -0,0 +1,365 @@
+        // block or the previous proof; it cannot be on a lower tier.
+        if (proof.tier == 0 || proof.tier < meta.minTier || proof.tier < ts.tier) {
+            revert L1_INVALID_TIER();
+        }
+
+        // Retrieve the tier configurations. If the tier is not supported, the
+        // subsequent action will result in a revert.
+        ITierProvider.Tier memory tier =
+            ITierProvider(resolver.resolve("tier_provider", false)).getTier(proof.tier);
+
+        _checkProverPermission(blk, ts, tid, tier);
+
+        // We must verify the proof, and any failure in proof verification will
+        // result in a revert.
+        //
+        // It's crucial to emphasize that the proof can be assessed in two
+        // potential modes: "proving mode" and "contesting mode." However, the
+        // precise verification logic is defined within each tier'IVerifier
+        // contract implementation. We simply specify to the verifier contract
+        // which mode it should utilize - if the new tier is higher than the
+        // previous tier, we employ the proving mode; otherwise, we employ the
+        // contesting mode (the new tier cannot be lower than the previous tier,
+        // this has been checked above).
+        //
+        // It's obvious that proof verification is entirely decoupled from
+        // Taiko's core protocol.
+        {
+            address verifier = resolver.resolve(tier.verifierName, true);
+            // The verifier can be address-zero, signifying that there are no
+            // proof checks for the tier. In practice, this only applies to
+            // optimistic proofs.
+            if (verifier != address(0)) {
+                bool isContesting = proof.tier == ts.tier && tier.contestBond != 0;
+
+                IVerifier.Context memory ctx = IVerifier.Context({
+                    metaHash: blk.metaHash,
+                    blobHash: meta.blobHash,
+                    prover: msg.sender,
+                    blockId: blk.blockId,
+                    isContesting: isContesting,
+                    blobUsed: meta.blobUsed
+                });
+
+                IVerifier(verifier).verifyProof(ctx, tran, proof);
+            }
+        }
+
+        IERC20 tko = IERC20(resolver.resolve("taiko_token", false));
+
+        // A special return value from the top tier prover can signal this
+        // contract to return all liveness bond.
+        {
+            bool returnLivenessBond = tier.contestBond == 0 && blk.livenessBond > 0
+                && proof.data.length == 32 && bytes32(proof.data) == RETURN_LIVENESS_BOND;
+
+            if (returnLivenessBond) {
+                tko.transfer(blk.assignedProver, blk.livenessBond);
+                blk.livenessBond = 0;
+            }
+        }
+
+        bool sameTransition = tran.blockHash == ts.blockHash && tran.signalRoot == ts.signalRoot;
+
+        if (proof.tier > ts.tier) {
```

Comment:
@adaki2004 I created this file to live side-by-side with the previous implementation. 

You said there may be a bug (lower tier can contest a higher tier proof), according to the new test test_L1_ContestingWithLowerTierProofRevertspn, it's not possible.